config/routes.rb

```diff
@@ -607,7 +607,8 @@
             put :suspend
             put :resume
           end
-          resources :keys, :controller => 'buyer_application_keys', :only => [:index, :create, :destroy]
+          resources :keys, constraints: { id: %r{([^/]+?)(?=\.json|\.xml|$|/)} }, controller: 'buyer_application_keys', only: %i[index create destroy]
```

Comment:
A few thoughts about the regexp:

1. AFAIK `([^/]+?)` means "accept the less possible ammount of anything expect a slash". But slashes are going to be accepted too, right? (https://github.com/3scale/porta/pull/3711)
2. This regexp would accept spaces, is that expected?
3. Why not just use the same regexp we use in the validator?

```suggestion
          resources :keys, constraints: { id: %r{([\x21-\x7E]+?)(?=\.json|\.xml|$|/)} }, controller: 'buyer_application_keys', only: %i[index create destroy]
```

Reply:
Thanks @jlledom 

1. Yes, the slashes are accepted, but they need to be URL-encoded, as well as other special characters :sunglasses: Dots, however, belong to [unreserved characters](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3) so they are not encoded, that's why we need to account for them explicitly. So, here we just care about 

2. Same as above, if a space would be a part of the application key, it would need to be URL-encoded by the API client.

3. I think that's giving too much responsibility to the router. I think while format extensions `.json` and `xml` can be considered part of the routing logic, the allowed characters, IMO, should definitely be part of the business logic.

P.S. Now having said all the above... I am not sure whether URL-decoding happens before the router logic, of after :upside_down_face: So, if it happens before, 1. and 2. are not anymore valid...

Reply:
UPDATE: I think routing happens before decoding because if I change `[^/%]` to `[^/%]` and send

```
curl -v -g -X DELETE 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/667/keys/hello{}.json?access_token=admintoken'
```

I am getting a routing error:
```
No route matches [DELETE] "/admin/api/accounts/3/applications/667/keys/hello%7B%7D.json"
```

(so it tries to match `%` and not `{`)

Also, in production routing errors are actually reported in a very ugly way... This HTML is returned: https://github.com/3scale/porta/blob/f1ce4b505ce139e229aa49e2979744f180147c02/public/404.html

Well, I guess that's because the logic is generic for the UI and API.

On the contrary, in case routing has passed but the key is missing the request will return `404` with empty body for XML, and `404` with `{"status":"Not found"}` for JSON, which is more user-friendly.

Reply:
I agree with @mayorova here.

Reply:
@mayorova Thanks for your comments, however, I'm so lost! I don't understand your example. Let's say the application key is actually `hello{}` on the DB. Then, you call:

```
curl -v -g -X DELETE 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/667/keys/hello{}.json?access_token=admintoken'
``` 

In this case `hello{}` is not encoded. Is it somehow encoded to `hello%7B%7D` at some moment, and you expect it to be decoded back to `hello{}` before trying to match the route?

Reply:
Hahaha @jlledom, now I'm lost too :sweat_smile:

I actually made a typo in my example, so it made it very confusing. But on the other hand, you uncovered a behavior which I hadn't noticed before :upside_down_face: 

So, first, the things that are clear:

While `hello{}` can be used in the URL path, it's not the right way to form a URL, as most special characters (except some unreserved and sub-delimiters chars) need to be percent-encoded, so the correct URL would be:
```
curl -v -g -X DELETE 'http://provider-admin.3scale.localhost:3000/admin/api/accounts/3/applications/667/keys/hello%7B%7D.json?access_token=admintoken'
```

For example, ActiveDocs does this percent-encoding:

![image](https://github.com/3scale/porta/assets/1270328/7a5e3695-b7ee-47b0-b1af-1c2dea21238a)

This is also what we do in [Pisoni](https://github.com/search?q=repo%3A3scale%2Fpisoni%20url_encode&type=code) before sending the request to apisonator, and apisonator in its turn decodes the path.

Now, let's return to the confusing things...

My doubt with Rails was whether routing happens before decoding or after.

My example should have been with `curl [...]/keys/hello%7B%7D.json[...]`

So, the reasoning is: if we prohibit `%` in the key IDs, this curl does not match, which means that it's not decoded before matching (otherwise, it would try to match `hello{}` and that would pass in theory).

However, I can now see that the results are exactly the same with both values, encoded or not :upside_down_face:  So, my guess is that the URL first gets "normalized" (which may include encoding), then matched, and then decoded to extract the parameter values. But I don't have proof yet :grimacing:

Reply:
So if I understood this correctly, It doesn't matter if you send `hello{}` or `hello%7B%7D`, in both cases it returns `No route matches`?

It that case, why the second one doesn't math the route?

Reply:
why don' we add validation and not allow dots in the key?

Reply:
@nidhi-soni1104 What are you referring to? We do have a validation for keys, which **allows** dots in keys. But this PR fixes a case where this causes a problem.

Reply:
I think this globbing thing is more tricky... But there shouldn't actually be a problem with slashes. If you have a slash in the key (e.g. `hell/o`), you would need to send it as `hell%2Fo`.

I couldn't figure out exactly where the decoding happens, but it does happen apparently :D